index.js

```diff
@@ -1,174 +1,221 @@
-      }
-      if (isBinary(this._decoded)) return copyBinary(this._decoded)
-      if (immutableTypes.has(typeof this._decoded) || this._decoded === null) {
-        return this._decoded
-      }
-      return clone(this._decoded)
+  _decode () {
+    if (typeof this.opts.source !== 'undefined') this._decoded = this.opts.source
+    else {
+      this._decoded = this.opts.codec.decode(this._encoded || this.opts.data)
     }
+    return this._decoded
+  }
 
-    decodeUnsafe () {
-      if (!this._decoded) this._decode()
+  decode () {
+    // TODO: once we upgrade to the latest data model version of
+    // dag-pb that @gozala wrote we should be able to remove this
+    // and treat it like every other codec.
+    /* c8 ignore next */
+    if (this.codec === 'dag-pb') return this._decode()
+    if (!this._decoded) this._decode()
+    const tt = typeof this._decoded
+    if (tt === 'number' || tt === 'boolean') {
+      // return any immutable types
       return this._decoded
     }
-
-    reader () {
-      return reader(this.decodeUnsafe())
+    if (isBinary(this._decoded)) return copyBinary(this._decoded)
+    if (immutableTypes.has(typeof this._decoded) || this._decoded === null) {
+      return this._decoded
     }
+    return clone(this._decoded)
+  }
 
-    async equals (block) {
-      if (block === this) return true
-      const cid = await this.cid()
-      if (CID.asCID(block)) return cid.equals(CID.asCID(block))
-      // https://github.com/bcoe/c8/issues/135
-      /* c8 ignore next */
-      return cid.equals(await block.cid())
-    }
+  decodeUnsafe () {
+    if (!this._decoded) this._decode()
+    return this._decoded
   }
 
-  const BlockWithIs = withIs(Block, { className: 'Block', symbolName: '@ipld/block' })
-  BlockWithIs.encoder = (source, codec, algo) => new BlockWithIs({ source, codec, algo })
-  BlockWithIs.decoder = (data, codec, algo) => new BlockWithIs({ data, codec, algo })
-  BlockWithIs.create = (data, cid) => {
-    if (typeof cid === 'string') cid = CID.from(cid)
-    return new BlockWithIs({ data, cid })
+  reader () {
+    return reader(this.decodeUnsafe())
+  }
+
+  async equals (block) {
+    if (block === this) return true
+    const cid = await this.cid()
+    if (block.asCID === block) return cid.equals(block)
+    // https://github.com/bcoe/c8/issues/135
+    /* c8 ignore next */
+    return cid.equals(await block.cid())
   }
-  BlockWithIs.multiformats = multiformats
-  BlockWithIs.CID = CID
-  return BlockWithIs
 }
 
-export default create
+Block.codecs = new Map()
+
+Block.add = codec => {
+  if (codec.name) Block.codecs.set(codec.name, codec)
+  if (codec.code) Block.codecs.set(codec.code, codec)
+}
+Block.add(json)
+Block.add(raw)
+
+Block.encoder = (source, codec, hasher) => new Block({ source, codec, hasher })
+Block.decoder = (data, codec, hasher) => new Block({ data, codec, hasher })
+Block.create = (data, cid) => {
+  if (typeof cid === 'string') cid = CID.parse(cid)
```

Comment:
💭 I would suggest to either remove support for `cid` in string representation or have a separate method for it. Because if `CID.parse` is not provided a base decoder it would throw unless it's in `base32` or `base58btc`.

It seems to me that this library should not try to concern itself with CID parsing, it users should do it if necessary.

Reply:
ya, i’m inclined to just drop support for strings.

Reply:
+1